lib/ethon/curls/functions.rb

```diff
@@ -14,14 +14,8 @@ def self.extended(base)
 
         base.attach_function :easy_init,                  :curl_easy_init,           [],                             :pointer
         base.attach_function :easy_cleanup,               :curl_easy_cleanup,        [:pointer],                     :void
-        base.attach_function :easy_getinfo,               :curl_easy_getinfo,        [:pointer, :info, :pointer],    :easy_code
-        base.attach_function :easy_setopt_ffipointer,     :curl_easy_setopt,         [:pointer, :easy_option, :pointer],  :easy_code
-        base.attach_function :easy_setopt_string,         :curl_easy_setopt,         [:pointer, :easy_option, :string],   :easy_code
-        base.attach_function :easy_setopt_long,           :curl_easy_setopt,         [:pointer, :easy_option, :long],     :easy_code
-        base.attach_function :easy_setopt_callback,       :curl_easy_setopt,         [:pointer, :easy_option, :callback], :easy_code
-        base.attach_function :easy_setopt_debug_callback, :curl_easy_setopt,         [:pointer, :easy_option, :debug_callback], :easy_code
-        base.attach_function :easy_setopt_progress_callback, :curl_easy_setopt,      [:pointer, :easy_option, :progress_callback], :easy_code
-        base.attach_function :easy_setopt_off_t,          :curl_easy_setopt,         [:pointer, :easy_option, :int64],    :easy_code
+        base.attach_function :easy_getinfo,               :curl_easy_getinfo,        [:pointer, :info, :varargs],    :easy_code
+        base.attach_function :easy_setopt,                :curl_easy_setopt,         [:pointer, :easy_option, :varargs], :easy_code
         base.instance_variable_set(:@blocking, true)
         base.attach_function :easy_perform,               :curl_easy_perform,        [:pointer],                     :easy_code
```

Comment:
I'm still failing on this function. 
Doesn't need any changes here?

```

/home/username/.rvm/gems/ruby-2.4.1/bundler/gems/ethon-90889ce772a9/lib/ethon/easy/operations.rb:44: [BUG] Segmentation fault at 0x007fac9d2d3e20
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]

-- Control frame information -----------------------------------------------
c:0028 p:---- s:0154 e:000153 CFUNC  :easy_perform
c:0027 p:0018 s:0149 e:000148 METHOD /home/username/.rvm/gems/ruby-2.4.1/bundler/gems/ethon-90889ce772a9/lib/ethon/easy/operations.rb:44
c:0026 p:0027 s:0145 e:000144 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/operations.rb:16
c:0025 p:0041 s:0140 e:000139 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/cacheable.rb:18
c:0024 p:0041 s:0135 e:000134 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/block_connection.rb:31
c:0023 p:0058 s:0131 e:000130 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/stubbable.rb:25
c:0022 p:0027 s:0126 e:000125 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/before.rb:26
c:0021 p:0030 s:0122 e:000121 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/typhoeus-1.4.0/lib/typhoeus/request/actions.rb:22
c:0020 p:0053 s:0116 e:000115 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/sources_manager.rb:76
c:0019 p:0020 s:0108 e:000107 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/sources_manager.rb:36
c:0018 p:0020 s:0097 e:000096 METHOD /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/sources_manager.rb:21
c:0017 p:0013 s:0092 e:000091 BLOCK  /home/username/.rvm/gems/ruby-2.4.1/gems/cocoapods-1.10.1/lib/cocoapods/installer/analyzer.rb:178 [FINISH]
```

Reply:
I am still getting an error here as well.

Reply:
In contrary to the other methods I fixed here, [`curl_easy_perform`](https://curl.se/libcurl/c/curl_easy_perform.html) 's declaration seems correct to me.
I do not have a Linux environment at hand but the tests do run the method that crashed on that stack trace, and the tests did pass on the Ubuntu CI 🤔
Also with or without my change I'm pretty sure ethon is working OK on x86_64 (your stack trace tells me you are running Ruby 2.4.1 on x64_64 in a Linux environment). Without my patch do you have any problem on that environment?

BTW Ruby 2.4 [should not be used anymore](https://www.ruby-lang.org/en/news/2020/04/05/support-of-ruby-2-4-has-ended/) (and  it will be the same for 2.5 very soon), even though I does not look like to be the cause for the crash here.

Reply:
Yes, the same errors without your patch.

Reply:
I get the same seg faults in the ethon "easy" methods, including  `curl_easy_perform`, with ruby 2.6.5 running on Alpine Linux with musl libc.

Reply:
So it appears that the errors on `curl_easy_perform` in Linux are possibly a similar but unrelated issue?

Meanwhile, this PR _does_ fix the segfault on arm64 macOS, which (among other things) is the only remaining blocker to running Cocoapods natively on an M1 Mac. I think it's fair to say this is something a lot of people really want to be able to do.

If you compare the output of the latest failed (truffleruby-head, macOS) test run here with the output from a [recent successful run of it](https://github.com/typhoeus/ethon/pull/179/checks?check_run_id=1947340458), you'll see that the tests did pass for this, but the test harness threw an exception at the end over a thread that didn't converge. This suggests to me that the error was just flakiness in the test harness, and does not reflect any problem with this PR. Previous runs on this MR failed on (truffleruby-head, ubuntu) but passed on (truffleruby-head, macOS), which also suggests that it's flakiness with truffleruby-head and not this PR. This same failure also occurred against master a month ago: https://github.com/typhoeus/ethon/runs/1863459474?check_suite_focus=true

All of which is to say that I don't think there are any real blockers remaining to merging this, and given that it fixes a major crash bug affecting many users... could someone please review and hopefully approve this?